src/expr/src/relation/mod.rs

```diff
@@ -674,98 +674,124 @@ impl MirRelationExpr {
+                    for key_set in &mut input {
+                        key_set.retain(|k| !cols_equal_to_literal.contains(&k));
+                        for col in key_set.iter_mut() {
+                            if let Some(equiv) = union_find.get(*col) {
+                                *col = *equiv;
+                            }
+                        }
+                        key_set.sort();
+                        key_set.dedup();
+                    }
+                    input.sort();
+                    input.dedup();
+
+                    // Expand out each key to each of its equivalent forms.
+                    // Each instance of `col` can be replaced by any equivalent column.
+                    // This has the potential to result in exponentially sized number of unique keys,
+                    // and in the future we should probably maintain unique keys modulo equivalence.
+                    let mut subs = Vec::new();
+                    for (col, sub) in union_find.iter().enumerate() {
+                        if *sub != col {
+                            assert!(*sub < col);
+                            while subs.len() <= *sub {
+                                subs.push(Vec::new());
+                            }
+                            subs[*sub].push(col);
+                        }
+                    }
+                    // For each column, substitute for it in each occurrence.
+                    let mut to_add = Vec::new();
+                    for (col, subs) in subs.iter().enumerate() {
+                        if !subs.is_empty() {
+                            for key_set in input.iter() {
+                                if key_set.contains(&col) {
+                                    let mut to_extend = key_set
+                                        .iter()
+                                        .cloned()
+                                        .filter(|c| c != &col)
+                                        .collect::<Vec<_>>();
+                                    for sub in subs {
+                                        to_extend.push(*sub);
+                                        to_add.push(to_extend.clone());
+                                        to_extend.pop();
                                     }
-                                    new_keys.insert(key);
                                 }
                             }
-                            prev_keys = new_keys;
                         }
+                        // No deduplication, as we cannot introduce duplicates.
+                        input.append(&mut to_add);
                     }
-
-                    prev_keys.into_iter().sorted().collect()
-                } else {
-                    input
+                    for key_set in input.iter_mut() {
+                        key_set.sort();
+                        key_set.dedup();
+                    }
+                    input.sort();
```

Comment:
nit: this and the `input.dedup()` can be omitted as those are also done on the result-bound `keys`.